Handle CUPTI_ERROR_MAX_LIMIT_REACHED and disable TF tracing

kaixih opened the pull request:
This PR fixes an issue caused by multiple processes attempting to use CUPTI simultaneously. For example, when using NVIDIA Nsight System with TF tracing at the same time (both rely on CUPTI APIs), users will get wrong TF tracing and incomplete Nsight reports.

According to https://docs.nvidia.com/cupti/Cupti/r_limitations.html#r_limitations:

> The application which calls CUPTI APIs cannot be used with Nvidia tools like nvprof, Nvidia Visual Profiler, Nsight Compute, Nsight Systems, Nvidia Nsight Visual Studio Edition, cuda-gdb and cuda-memcheck.

So, we should avoid calling CUPTI APIs if the function cuptiSubscribe() detects CUPTI_ERROR_MAX_LIMIT_REACHED.

In this PR, we let TensorFlow issue a warning when that CUPTI_ERROR_MAX_LIMIT_REACHED is detected and skip all the following CUPTI calls. Then, the external profiling tool can have exclusive access to CUPTI APIs.

FYI. @nluehr

sanjoy commented:
CC @jbaiocchi (I tried assigning you as the reviewer but for some reason github does not let me)

sanjoy commented on tensorflow/core/profiler/internal/gpu/cupti_tracer.cc:
```diff
@@ -1380,32 +1381,47 @@ void CuptiTracer::Enable(const CuptiTracerOptions &option,
+      LOG(ERROR) << "function cupti_interface_->Subscribe failed with error "
+                 << errstr;
+    }
+    return errors::Internal(absl::StrCat("cupti call error", errstr));
```
I believe this can just be `errors::Internal("cupti call error", errstr)`.

Also, should there be a space after `"cupti call error"`?

trisolaran commented on tensorflow/core/profiler/internal/gpu/cupti_tracer.cc:
```diff
@@ -1380,32 +1381,47 @@ void CuptiTracer::Enable(const CuptiTracerOptions &option,
 
   VLOG(1) << "Enable subscriber";
-  RETURN_IF_CUPTI_ERROR(cupti_interface_->Subscribe(
-      &subscriber_, (CUpti_CallbackFunc)ApiCallback, this));
```
cannot you just move api_tracing_enabled_ = true;  after this?

trisolaran commented on tensorflow/core/profiler/internal/gpu/cupti_tracer.cc:
```diff
@@ -1380,32 +1381,47 @@ void CuptiTracer::Enable(const CuptiTracerOptions &option,
   }
 
   EnableApiTracing().IgnoreError();
```
and don't IgnoreError() here.

yisitu commented:
Unfortunately supporting external tools concurrently with TensorFlow tracing and profiler is not a use case that we support. Our tools are not designed to work concurrently with NVIDIA tools, hence this is the expected behavior.

It would help to explain the problem you are trying to solve, that would involve requiring Nsight Systems?

houtoms commented:
@yisitu This PR's purpose is to support the external profiling tools and at the same time not to affect existing TF tracing.

One use case: suppose the example.py has TF tracing in it. If we do `nsys profile python example.py`, both Nsys tracing and TF tracing will collect nothing. With this change, TF tracing will be ignored and Nsys gets correct report. And `python example.py` won't be affected.

trisolaran commented:
I know that in your case, once subscribe exist, it probably lasts for the life time of the process. however  there might be other cases that subscriber can collide with each other,  e.g. on-demand with tensorboard callback, we don't want to disable it permanently, so the next call to Enable() (when all other subscribers have expired) should be successful, 
how do you think?

yisitu commented:
In your use case, could you have disabled TF tracing in the first place
since you will be using `nsys profile`?

On Wed, Mar 18, 2020 at 11:20 AM Jie Sun <notifications@github.com> wrote:

> I know that in your case, once subscribe exist, it probably lasts for the
> life time of the process. however there might be other cases that
> subscriber can collide with each other, e.g. on-demand with tensorboard
> callback, we don't want to disable it permanently, so the next call to
> Enable() (when all other subscribers have expired) should be successful,
> how do you think?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/37674#issuecomment-600787844>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AONQZYDG5A66HH7KDS7APULRIEGGDANCNFSM4LN6NOMA>
> .
>

houtoms commented:
@yisitu Do you mean by changing the python script?

yisitu commented:
Yes, I imagine that the python script was instrumented to enable profiling
in the first place.

On Wed, Mar 18, 2020 at 11:39 AM Kaixi Hou <notifications@github.com> wrote:

> @yisitu <https://github.com/yisitu> Do you mean by changing the python
> script?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/37674#issuecomment-600796634>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AONQZYHPOZDH3C6PB6X6RNTRIEIMPANCNFSM4LN6NOMA>
> .
>

houtoms commented:
@trisolaran Yes, can you confirm my understanding of your example is correct? So, before this change, TF can handle the case that issues multiple subscriber sessions: if subscriber exists, ignore it; if not exist, create it. However, after this change, only the first session can create the subscriber but all the other sessions cannot create new subscriber anymore.

trisolaran commented:
yes. I think that you statement is right.
 I wonder if we can fail silently for this time, and able to enable cupti tracer next time. ( not applicable to your case though).

houtoms commented:
I am wondering adding more conditions in the Disable() might solve the problem. Basically, we need to make sure we still unsubscribe if we know the existing subscriber is from TF not external tools. Something like (isSet() is an imaginary function for now.)
```c++
void CuptiTracer::Disable() {
  if (!subscriber_existed || (subscriber_existed && subscriber_.isSet())) {
   //do as normal
  }
}
```

gbaned commented:
@houtoms Can you please resolve conflicts? Thanks!

tensorflowbutler commented:
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

gbaned commented:
@houtoms Any update on this PR, please. Thanks!

kaixih committed:
Fix TF cupti multi subscribers issue

Disable all other profilers if ext profiler found

Changed function names and update comments

Change warning sentence

Modified error code

houtoms commented:
Hi, I rebased the code to include a new design to disable all TF tracers when the external profiler is detected.

kaixih commented:
It should work. I think you are referring changes here: https://github.com/tensorflow/tensorflow/commit/b97cc6fd437e3ab8b788853558733f9420125245, which I feel is similar to the original design that disables the cupti tracer then external profiler is detected. But, as we discussed previously, my understanding is that we'd like to turn off all TF tracers if external profiler is detected. That's why the new commit here is to put the detection code to an earlier phase of the profiler initialization session. Anyway, if you are OK with the original plan, we can close this one.

trisolaran commented:
thanks Kaixi.
